### PR TITLE
add simple stack depth limit mechanism

### PR DESCRIPTION
**Pull request by deoxxa**

### CODE REVIEW

File: runtime.go [hunk 1]
```diff
@@ -63,6 +64,14 @@ type _runtime struct {
 
 func (self *_runtime) enterScope(scope *_scope) {
 	scope.outer = self.scope
+	if self.scope != nil {
+		scope.depth = self.scope.depth + 1
+	}
+
+	if self.stackLimit != 0 && scope.depth >= self.stackLimit {
+		panic(self.panicRangeError("RangeError: Maximum call stack size exceeded"))
```
Comment (stevenh) [inline runtime.go]: RangeError knows it's a RangeError so we shouldn't need the message prefix here.

I can see you've followed some other examples of its use, but I believe they are wrong too.

Comment (deoxxa) [inline runtime.go]: I did think it was a bit fishy. I'm happy to change this, and perhaps the others as well, so long as they're still identifiable as RangeError to the casual observer of a JS stack trace.

Comment (deoxxa) [inline runtime.go]:
```
> (1).toString(40)
RangeError: RangeError: toString() radix must be between 2 and 36
    at toString (<native code>)
    at <anonymous>:1:2
```

Looks like you're correct - I'll fix this instance and put together another PR to fix the others.

**Comment by stevenh** [comment]
I see this increments scope.depth on enterScope but doesn't decrement it on leaveScope, which seems odd; is this intended?

**Comment by deoxxa** [comment]
Yep - the scope is unwrapped in `leaveScope`. This is actually setting the stack depth on the new child scope, which is about to become the scope for that particular runtime.

File: otto_test.go [hunk 2]
```diff
@@ -1650,6 +1650,56 @@ func Test_objectLength(t *testing.T) {
...
+	})
+
+	tt(t, func() {
+		_, vm := test()
+
+		vm.vm.SetStackDepthLimit(6)
+
+		_, err := vm.Run(code)
+
+		is(err == nil, true)
```
Comment (stevenh) [inline otto_test.go]: Looking at the test code you only have 6 levels 1 initial and 5 functions, so I would expect this not to error i.e. its a max which can be reached.

File: otto.go [hunk 3]
```diff
@@ -367,6 +367,10 @@ func (self Otto) SetRandomSource(fn func() float64) {
 	self.runtime.random = fn
 }
 
+func (self Otto) SetStackDepthLimit(limit int) {
```
Comment (stevenh) [inline otto.go]: Needs a comment for the method.

File: runtime.go [hunk 4]
```diff
@@ -63,6 +64,14 @@ type _runtime struct {
 
 func (self *_runtime) enterScope(scope *_scope) {
 	scope.outer = self.scope
+	if self.scope != nil {
+		scope.depth = self.scope.depth + 1
+	}
+
+	if self.stackLimit != 0 && scope.depth >= self.stackLimit {
```
Comment (stevenh) [inline runtime.go]: As an optimisation would it make sense to not do this check if the above if failed?

File: otto_test.go (hunk 2 above)
Comment (deoxxa) [inline otto_test.go]: Sure, that works for me.

Comment (deoxxa) [inline otto_test.go]: Wait, no - that's exactly what's happening. It's making sure `err` is nil. I need to get more familiar with the way this whole `tt` thing works, as it's a little bit unclear.

Comment (deoxxa) [inline otto_test.go]: I've added some comments on the tests to make it a bit easier to see what's happening.

**Pull request merged**